classes/gateways/class.pmprogateway_stripe.php

```diff
@@ -1661,88 +1721,160 @@ function getCustomer( &$order = false, $force = false ) {
+			global $pmpro_stripe_customer_id;
+			$pmpro_stripe_customer_id = $customer->id;
+			if ( ! function_exists( 'pmpro_user_register_stripe_customerid' ) ) {
+				function pmpro_user_register_stripe_customerid( $user_id ) {
+					global $pmpro_stripe_customer_id;
+					update_user_meta( $user_id, "pmpro_stripe_customerid", $pmpro_stripe_customer_id );
+				}
+				add_action( "user_register", "pmpro_user_register_stripe_customerid" );
 			}
 		}
 
-		// No customer id and user is paying, create one.
-		if ( ! empty( $order->payment_method_id ) ) {
-			try {
-				$this->customer = Stripe_Customer::create( array(
-					"description" => $name . " (" . $email . ")",
-					"email"       => $order->Email,
-				) );
-			} catch ( \Stripe\Error $e ) {
-				$order->error      = __( "Error creating customer record with Stripe:", 'paid-memberships-pro' ) . " " . $e->getMessage();
-				$order->shorterror = $order->error;
+		/**
+		 * Fires after a Stripe_Customer is created at checkout.
+		 *
+		 * @since Unknown
+		 * @deprecated TBD. Use pmpro_stripe_update_customer_from_user or pmpro_stripe_update_customer_from_order.
+		 *
+		 * @param array       $customer_args to be sent.
+		 * @param MemberOrder $order being used to create/update customer.
+		 */
+		do_action( 'pmpro_stripe_create_customer', $customer );
+		return $customer;
+	}
 
-				return false;
-			} catch ( \Throwable $e ) {
-				$order->error      = __( "Error creating customer record with Stripe:", 'paid-memberships-pro' ) . " " . $e->getMessage();
-				$order->shorterror = $order->error;
+	/**
+	 * Create/Update Stripe customer for a user.
+	 *
+	 * @since TBD
+	 *
+	 * @param int $user_id to create/update Stripe customer for.
+	 * @return Stripe_Customer|false
+	 */
+	function update_customer_from_user( $user_id ) {
+		$user = get_userdata( $user_id );
 
-				return false;
-			} catch ( \Exception $e ) {
-				$order->error      = __( "Error creating customer record with Stripe:", 'paid-memberships-pro' ) . " " . $e->getMessage();
-				$order->shorterror = $order->error;
+		if ( empty( $user->ID ) ) {
+			// User does not exist.
+			return false;
+		}
 
-				return false;
-			}
+		// Get the name for the customer.
+		$name = trim( $user->first_name . " " . $user->last_name );
+		if ( empty( $name ) ) {
+			// In case first and last names aren't set.
+			$name = $user->user_login;
+		}
 
-			if ( ! empty( $user_id ) ) {
-				// User logged in/etc.
-				update_user_meta( $user_id, "pmpro_stripe_customerid", $this->customer->id );
-			} else {
-				//user not registered yet, queue it up
-				global $pmpro_stripe_customer_id;
-				$pmpro_stripe_customer_id = $this->customer->id;
-				if ( ! function_exists( 'pmpro_user_register_stripe_customerid' ) ) {
-					function pmpro_user_register_stripe_customerid( $user_id ) {
-						global $pmpro_stripe_customer_id;
-						update_user_meta( $user_id, "pmpro_stripe_customerid", $pmpro_stripe_customer_id );
-					}
+		// Get data to update customer with.
```

Comment:
Let's check if the customer at Stripe already has a billing address, and if so let's not push the billing address we have in user meta to Stripe.

We should bring the `$customer = $this->get_customer_for_user( $user_id );` line from below above this so we can check the values on the customer.

Reply:
Alternatively, let's continue to consider fully synchronizing the customer billing address. If we use a webhook to update the billing address user meta when the address is updated at Stripe, then we're less likely to overwrite something we don't want to. We would need to handle loops.

Reply:
Applied this suggestion in commit aba4a58, fully synchronizing the customer billing address seems like something that we should keep in mind and talk about across the entirety of PMPro